pallets/cash/src/chains.rs

```diff
@@ -348,72 +394,138 @@ impl FromStr for ChainId {
     }
 
+    fn map_events_eth_like<F>(
+        block: &EthereumBlock,
+        mut chain_block_event_fn: F,
+    ) -> impl Iterator<Item = ChainBlockEvent> + '_
+    where
+        F: FnMut(ChainBlockNumber, EthereumEvent) -> ChainBlockEvent + 'static,
+    {
+        block
+            .events
+            .iter()
+            .map(move |e| chain_block_event_fn(block.number, e.clone()))
+    }
+
     pub fn events(&self) -> impl Iterator<Item = ChainBlockEvent> + '_ {
-        match self {
-            ChainBlock::Eth(block) => block
-                .events
-                .iter()
-                .map(move |e| ChainBlockEvent::Eth(block.number, e.clone())),
-        }
+        let return_value: Box<dyn Iterator<Item = ChainBlockEvent>> = match self {
+            ChainBlock::Eth(block) => {
+                Box::new(Self::map_events_eth_like(block, ChainBlockEvent::Eth))
+            }
+            ChainBlock::Matic(block) => {
+                Box::new(Self::map_events_eth_like(block, ChainBlockEvent::Matic))
+            }
+        };
+
+        return_value
     }
 }
 
 /// Type for describing a set of blocks coming from an underlying chain.
 #[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug, Types)]
 pub enum ChainBlocks {
     Eth(Vec<<Ethereum as Chain>::Block>),
+    Matic(Vec<<Polygon as Chain>::Block>),
 }
 
 impl ChainBlocks {
     pub fn chain_id(&self) -> ChainId {
         match self {
             ChainBlocks::Eth(_) => ChainId::Eth,
+            ChainBlocks::Matic(_) => ChainId::Matic,
         }
     }
 
     pub fn len(&self) -> usize {
         match self {
             ChainBlocks::Eth(blocks) => blocks.len(),
+            ChainBlocks::Matic(blocks) => blocks.len(),
         }
     }
 
+    fn blocks_eth_like<F>(
+        blocks: &Vec<EthereumBlock>,
+        mut chain_block_fn: F,
+    ) -> impl Iterator<Item = ChainBlock> + '_
+    where
+        F: FnMut(EthereumBlock) -> ChainBlock + 'static,
+    {
+        blocks.iter().map(move |b| chain_block_fn(b.clone()))
+    }
+
     pub fn blocks(&self) -> impl Iterator<Item = ChainBlock> + '_ {
-        match self {
-            ChainBlocks::Eth(blocks) => blocks.iter().map(|b| ChainBlock::Eth(b.clone())),
-        }
+        let return_value: Box<dyn Iterator<Item = ChainBlock>> = match self {
+            ChainBlocks::Eth(blocks) => Box::new(Self::blocks_eth_like(blocks, ChainBlock::Eth)),
+            ChainBlocks::Matic(blocks) => {
+                Box::new(Self::blocks_eth_like(blocks, ChainBlock::Matic))
+            }
+        };
+
+        return_value
+    }
+
+    fn blocks_numbers_eth_like(blocks: &Vec<EthereumBlock>) -> impl Iterator<Item = u64> + '_ {
+        blocks.iter().map(|b| b.number)
     }
 
     pub fn block_numbers(&self) -> impl Iterator<Item = u64> + '_ {
-        match self {
-            ChainBlocks::Eth(blocks) => blocks.iter().map(|b| b.number),
-        }
+        let return_value: Box<dyn Iterator<Item = u64>> = match self {
```

Comment:
Hm actually surprised it even complains about this one since they are always `u64` items

Reply:
yeah there is a reallly strange error on that one, that is actually the case with a lot of these where the item is actually the same exact type all the way down `EthereumBlock` for example, I couldn't really make sense of the explanation at the time from what i remember. Honestly though this is a very small allocation relative to an entire Vec. Although, they're all quite small allocations overall I guess.. I suppose it is possible that some long lists of blocks could overflow the stack though